File: algomonster/miscellaneous/word_search_2.py
from typing import List

class Trie:
    def __init__(self, val):
        self.val = val
        self.next = {}
        self.word = -1

    def insert(self, s, index, i=0):
        if i == len(s):
            self.word = index
            return
        self.next.setdefault(s[i], Trie(s[i]))
        self.next[s[i]].insert(s, index, i+1)

def get_directions(matrix, i, j):
    if i == -1:
        yield 0, 0
        return
    h = len(matrix)
    w = len(matrix[0])
    moves = [(i+1, j), (i, j+1), (i-1, j), (i, j-1)]
    for x, y in moves:
        if 0 <= x < h and  0 <= y < w:
            yield x, y

def word_search_ii(matrix: List[str], words: List[str]) -> List[str]:
    trie = Trie('$')
    found = []
    for i, word in enumerate(words):
        found.append(False)
        trie.insert(word, i)

    def dfs(matrix, i, j, word, visited, cur):
        if cur.word >= 0:
            found[cur.word] = True
            cur.word = -1
        for x, y in get_directions(matrix, i, j):
            if not visited[x][y] and matrix[x][y] in cur.next:
                visited[x][y] = True
                dfs(matrix, x, y, word + matrix[x][y], visited, cur.next[matrix[x][y]])
                visited[x][y] = False

    h = len(matrix)
    w = len(matrix[0])
    visited = [[False] * w for i in range(h)]
    for i in range(h):
        for j in range(w):
            if matrix[i][j] in trie.next:
                visited[i][j] = True
                dfs(matrix, i, j, matrix[i][j], visited, trie.next[matrix[i][j]])
                visited[i][j] = False
    
    res = []
    for i in range(len(found)):
        if found[i]:
            res.append(words[i])
    return res

if __name__ == '__main__':
    matrix = input().split()
    words = input().split()
    res = word_search_ii(matrix, words)
    print(' '.join(res))

File: leetcode/trie/WordSearch2.cpp
class Trie {
    public:
        unordered_map<char, Trie> next;
        int word_index = -1;

        void insert(string s, int i, int word_index) {
            if (i == s.length()) {
                this->word_index = word_index;
                return;
            }
            if (!this->next.count(s[i])) this->next[s[i]] = Trie();
            this->next[s[i]].insert(s, i+1, word_index);
        }
};

class Solution {
public:
    vector<pair<int, int>> get_directions(vector<string>& matrix, int i, int j) {
        vector<pair<int, int>> dirs {make_pair(i+1, j), make_pair(i, j+1), make_pair(i-1, j), make_pair(i, j-1)};
        vector<pair<int, int>> possible_dirs;
        for (pair<int, int> dir: dirs) {
            if (0 <= dir.first && dir.first < matrix.size() && 0 <= dir.second && dir.second < matrix[0].size()) {
                possible_dirs.push_back(dir);
            }
        }
        return possible_dirs;
    }

    void dfs(int i, int j, Trie* cur, string word, vector<string>& matrix, vector<bool>& found) {
        if (cur->word_index > -1) {
            found[cur->word_index] = true;
            cur->word_index = -1;
        }
        for(pair<int, int> dir: get_directions(matrix, i, j)) {
            if (matrix[dir.first][dir.second] != '&' && cur->next.count(matrix[dir.first][dir.second])) {
                char prev = matrix[dir.first][dir.second];
                matrix[dir.first][dir.second] = '&';
                dfs(dir.first, dir.second, &(cur->next[prev]), word + prev, matrix, found);
                matrix[dir.first][dir.second] = prev;
            }
        }
    }

    vector<string> findWords(vector<vector<char>>& matrix, vector<string>& words) {
        // add words to trie
        Trie trie = Trie();
        for (int i=0; i < words.size(); i++) {
            trie.insert(words[i], 0, i);
        }

        vector<bool> found(words.size(), false);
        Trie cur = trie;
        for (int i = 0; i < matrix.size(); i++) {
            for (int j = 0; j < matrix[0].size(); j++) {
                if (trie.next.count(matrix[i][j])) {
                    string s = "";
                    char prev = matrix[i][j];
                    matrix[i][j] = '&';
                    dfs(i, j, &(trie.next[prev]), s + prev, matrix, found);
                    matrix[i][j] = prev;
                }
            }
        }

        vector<string> res;
        for (int i=0; i < found.size(); i++) {
            if (found[i]) {
                res.push_back(words[i]);
            }
        }

        return res;
    }
};
